File: src/codes/june_challenge/day7/coinChange2.js
/**
 * @param {number} amount
 * @param {number[]} coins
 * @return {number}
 */
var change = function(amount, coins) {
    let dp = new Array(amount + 1).fill(0);
    dp[0] = 1;
    for (let c of coins) {
        for (let i = c; i <= amount; i++) {
            dp[i] += dp[i - c];
        }
    }
    return dp[amount];
};

File: src/codes/june_challenge/day7/CoinChange2.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int change(int amount, vector<int>& coins) {
        int row = coins.size() + 1;
        int col = amount + 1;
        vector<vector<int>> dp(row, vector<int>(col, 0));
        for (int i = 1; i < col; i++) {
            dp[0][i] = 0;
        }
        for (int i = 0; i < row; i++) {
            dp[i][0] = 1;
        }
        for (int i = 1; i < row; i++) {
            for (int j = 1; j < col; j++) {
                int temp = j - coins[i - 1];
                if(temp >= 0) {
                    dp[i][j] = dp[i][temp] + dp[i - 1][j];
                } else {
                    dp[i][j] = dp[i - 1][j];
                }
            }
        }
        return dp[row - 1][col - 1];
    }
};